clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -198,6 +208,41 @@ public boolean hasCompletedFetches() {
         return !completedFetches.isEmpty();
     }
 
+    /**
+     * Return if there are buffered CompletedFetches for a paused TopicPartition.
+     *
+     * @param tp {@link TopicPartition}
+     * @return  {@code true} if there are buffered {@link CompletedFetch},  {@false} otherwise
+     */
+    public boolean hasPausedCompletedFetchesFor(TopicPartition tp) {
+        return pausedCompletedFetchesPerTopicPartition.containsKey(tp);
+    }
+
+    /**
+     * Add the TopicPartition to the list of recently unpaused partitions.
+     * @param tp {@link TopicPartition}
+     */
+    public void addRecentlyUnpausedPartition(TopicPartition tp) {
+        this.recentlyUnPausedTopicPartitions.add(tp);
+    }
+
+    /**
+     * Remove the TopicPartition from list of recently unpaused partitions.
+     * @param tp {@link TopicPartition}
+     */
+    public void removeRecentlyUnpausedPartition(TopicPartition tp) {
```

Comment:
Can we simplify the patch by not having this method? We pause a partition after putting it in `recentlyUnPausedTopicPartitions`, we can still use `fetcher.subscriptions` to check whether the partition is unpaused while going over the partition in `recentlyUnPausedTopicPartitions`.

Reply:
Hi @lindong28 , can you shed more light on "We pause a partition after putting it in recentlyUnPausedTopicPartitions".
recentlyUnPausedTopicPartitions is used to give priority for returning buffered data for partitions that were paused earlier. When a partition is UnPaused, we add it to the "recentlyUnPausedTopicPartitions". When we Pause a partition, we remove it from "recentlyUnPausedPartitions". 
If we have a method for adding to recentlyUnPausedPartitions, we should have a method for removing it as well, right? 
The way I understand your suggested approach, if we want to use fetcher.subscriptions, we will have to iterate over all the fetchablePartitions in fetchedRecords() till we find the TopicPartition that is present in recentlyUnPausedPartitions and then check if there is data buffered for it. Iterating over fetcher.subscriptions, seems like an O(n) operation whereas using the recentlyUnpausedTopicPartitions as here, seems like O(m) operation where m < n.

Reply:
Regarding `If we have a method for adding to recentlyUnPausedPartitions, we should have a method for removing it as well, right?`, I don't think there is strict rule that says we must have a public API to remove it, as long as this can be removed. Currently `fetchedRecords()` will actually remove entry from `recentlyUnPausedPartitions` if `!pausedCompletedFetchesPerTopicPartition.containsKey(tp) && !pausedNextInLineRecordsPerTopicPartition.containsKey(tp)` is evaluated to true.

I understand that we need `recentlyUnPausedPartitions` for improving the time complexity as you mentioned. Here my suggestion is that we don't need the method `removeRecentlyUnpausedPartition(TopicPartition tp)`. So question here is that, is there performance or correctness concern if we don't have `removeRecentlyUnpausedPartition(...)`?

Reply:
Right, the alternative solution in the code snippet above is what I am suggesting. So we agree that we can achieve the same correctness and performance without having method `removeRecentlyUnpausedPartition`.

All things being equal, it is always preferred to have less API (and logic) exposed from Fetcher.java. So the question here is, what is the benefit of having this extra API. I am just not convinced by the statement that this method makes code more explicit and readable. Whether all not this is more readable is an objective opinion and may be you for opinion from other developers.

Reply:
Hi @lindong28, I am not 100% convinced with "it is always preferred to have less API (and logic) exposed from Fetcher.java". This class is internal to the KafkaConsumer logic and not exposed by an api directly from KafkaConsumer. We already have more powerful public apis like : sendFetches(), getTopicMetadata(), fetchedRecords() exposed in this class as public methods. Also we are publicly exposing an api to add to `recentlyUnPausedPartitions`. So not having a method to remove from it, seems odd to me.
I agree with "Whether all not this is more readable is an objective opinion". We can wait for someone else to take a look at the discussion here and let us know there opinions. Ping :  @jjkoshy @junrao @onurkaraman